build/builder_test.go

```diff
@@ -230,3 +234,268 @@ func TestDontCountContentOfSkippedFiles(t *testing.T) {
 		t.Fatalf("content of skipped documents should not count towards shard size thresold")
 	}
 }
+
+func TestOptions_FindAllShards(t *testing.T) {
+	type normalShard struct {
+		Repository zoekt.Repository
+		// NumShards is the number of shards that should be created that
+		// contain data for "Repository".
+		NumShards int
+	}
+
+	tests := []struct {
+		name               string
+		skipReason         string
+		normalShards       []normalShard
+		compoundShards     [][]zoekt.Repository
+		expectedShardCount int
+		expectedRepository zoekt.Repository
+	}{
+		{
+			name: "repository in normal shard",
+			normalShards: []normalShard{
+				{Repository: zoekt.Repository{Name: "repoA", ID: 1}},
+				{Repository: zoekt.Repository{Name: "repoB", ID: 2}},
+				{Repository: zoekt.Repository{Name: "repoC", ID: 3}},
+			},
+			expectedShardCount: 1,
+			expectedRepository: zoekt.Repository{Name: "repoB", ID: 2},
+		},
+		{
+			name: "repository in compound shard",
+			compoundShards: [][]zoekt.Repository{
+				{
+					{Name: "repoA", ID: 1},
+					{Name: "repoB", ID: 2},
+					{Name: "repoC", ID: 3},
+				},
+				{
+					{Name: "repoD", ID: 4},
+					{Name: "repoE", ID: 5},
+					{Name: "repoF", ID: 6},
+				},
+			},
+			expectedShardCount: 1,
+			expectedRepository: zoekt.Repository{Name: "repoB", ID: 2},
+		},
+		{
+			name: "repository split across multiple shards",
+			normalShards: []normalShard{
+				{Repository: zoekt.Repository{Name: "repoA", ID: 1}},
+				{Repository: zoekt.Repository{Name: "repoB", ID: 2}, NumShards: 2},
+				{Repository: zoekt.Repository{Name: "repoC", ID: 3}},
+			},
+			expectedShardCount: 2,
+			expectedRepository: zoekt.Repository{Name: "repoB", ID: 2},
+		},
+		{
+			name: "unknown repository",
+			normalShards: []normalShard{
+				{Repository: zoekt.Repository{Name: "repoA", ID: 1}},
+				{Repository: zoekt.Repository{Name: "repoB", ID: 2}},
+				{Repository: zoekt.Repository{Name: "repoC", ID: 3}},
+			},
+			compoundShards: [][]zoekt.Repository{
+				{
+					{Name: "repoD", ID: 4},
+					{Name: "repoE", ID: 5},
+					{Name: "repoF", ID: 6},
+				},
+			},
+			expectedShardCount: 0,
+		},
+		{
+			name: "match on ID, not name (compound only)",
+			compoundShards: [][]zoekt.Repository{
+				{
+					{Name: "repoA", ID: 1},
+					{Name: "sameName", ID: 2},
+					{Name: "sameName", ID: 3},
+				},
+				{
+					{Name: "repoB", ID: 4},
+					{Name: "sameName", ID: 5},
+					{Name: "sameName", ID: 6},
+				},
+			},
+			expectedShardCount: 1,
+			expectedRepository: zoekt.Repository{Name: "sameName", ID: 5},
+		},
+		{
+			name:       "match on ID, not name",
+			skipReason: "We only support matching on IDS for compound shards. For normal shards, repositories wih the same name will produce shards with the same filename (which breaks our matching logic).",
```

Comment:
It took me a bit to realize what was going on here. 

Our builder logic creates shards with filenames that are based off the repository name. There are two failure modes that I can think of:
- If we're asked to index two repositories with the same name, but different ids (ex: `{Name: "sameName", ID: 1},{Name: "sameName", ID: 2}`), the shard they'll create will have the same filename: `sameName.00000.zoekt`. This means that they'll clobber each other - both indexes can't exist at the same time on disk.
- If a repository is renamed to be the same as an existing one (ex: `{Name: "repoA", ID: 1} -> {Name: "repoB", ID: 1}`, and we already have `{Name: "repoB", ID: 2}` indexed), then the shard name for both repositories will be the same `repoB.00000.zoekt`. Because of this, calling FindAllShards on for repo ID `1` will return shards intended for repo ID `2`. This can also happen if `repoA` and `repoB` swap names. 

These are probably niche cases, but ideally we'd want to make our shard names based off of unique identifiers instead of just names. If we can't change the of the shards, then the only way that I can think of enforcing correctness is to read the shard's metadata by calling zoket.ReadMetadata(...). I think this would be much more expensive than just a `stat` syscall. 

---

I'm happy to delete this test instead of skipping it, but I wanted to leave it in for discussion.

Reply:
We rely on the name of simple shards being unique because it derives from the URI code-host/path/name which we trust to be unique. We should move to IDs at some point, although that will make dev and debugging a bit harder.

I am not sure it is fair to say "We only suuport matching IDS for compound shards". We can match simple shards by IDs too. Rather we don't support indexing of repos with the same name, because the shards end up overwriting each other.